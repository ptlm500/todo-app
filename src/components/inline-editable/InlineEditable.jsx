import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export default function InlineEditable(props) {
  const {
    className,
    value,
    placeholder,
    children,
    onSubmit
  } = props;

  if (Array.isArray(children)) {
    throw new Error('[InlineEditable]: Multiple child components passed. InlineEditable only accepts one child', children);
  }

  const childRef = useRef();

  const [editing, setEditing] = useState(false);

  useEffect(() => {
    if (editing && childRef && childRef.current) {
      childRef.current.focus();
    }
  }, [editing, childRef]);

  const childWithProps = React.cloneElement(children, {
    ref: childRef,
    value,
    placeholder
  });

  function handleKeyDown(e) {
    if (e.key === 'Escape') {
      finishEdit();
    }
  }

  function finishEdit() {
    setEditing(false);
    onSubmit && onSubmit(value);
  }

  return (
    <div
      className={className ? `${className} inline-editable` : 'inline-editable'}
    >
      {
        editing ? (
          <div
            className='inline-editable__edit-container'
            onBlur={finishEdit}
            onKeyDown={handleKeyDown}
          >
            {childWithProps}
          </div>
        ) : (
          <div
            className='inline-editable__edit-trigger'
            onClick={() => setEditing(true)}
          >
            <Value>
              {value || placeholder || 'Provide a placeholder value'}
            </Value>
          </div>
        )
      }
    </div>
  );
}

const Value = styled.span`
  white-space: pre-wrap;
  word-break: break-word;
`;

InlineEditable.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  value: PropTypes.string,
  placeholder: PropTypes.string,
  onSubmit: PropTypes.func
};
