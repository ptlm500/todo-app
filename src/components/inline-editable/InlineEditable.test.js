import React from 'react';
import renderer, { act } from 'react-test-renderer';
import InlineEditable from './InlineEditable';

const mockFocus = jest.fn();
const mockOnSubmit = jest.fn();

const defaultProps = {
  value: 'test value',
  placeholder: 'test placeholder',
  onSubmit: mockOnSubmit
};

function testRender(props = defaultProps, children) {
  let tree;

  act(() => {
    tree = renderer.create(
      <InlineEditable {...props}>
        {children ? children : <input />}
      </InlineEditable>,
      {
        createNodeMock: element =>
          element.type === 'input' && {
            focus: mockFocus
          }
      }
    );
  });

  return tree;
}

beforeEach(() => {
  jest.clearAllMocks();
});

describe('InlineEditable', () => {
  it('throws an error when multiple child components are passed to it', () => {
    expect.assertions(1);
    try {
      act(() =>
        renderer.create(
          <InlineEditable>
            <input/>
            <input/>
          </InlineEditable>
        )
      );
    } catch (e) {
      expect(e.message).toMatchSnapshot();
    }
  });

  it('renders a value when provided', () => {
    expect(testRender()).toMatchSnapshot();
  });

  it('sets a className on the outer div when provided', () => {
    const props = {
      className: 'test-class',
      ...defaultProps
    };

    expect(testRender(props)).toMatchSnapshot();
  })

  it('renders a placeholder when no value is provided', () => {
    expect(testRender({placeholder: 'test placeholder'})).toMatchSnapshot();
  });

  it('renders a message to provide a placeholder when no value or placeholder is provided', () => {
    expect(testRender({})).toMatchSnapshot();
  });

  describe('edit mode', () => {
    function enterEditMode(testRenderer) {
      const editTrigger = testRenderer.root
        .findByProps({className: 'inline-editable__edit-trigger'});

      act(() => editTrigger.props.onClick());
    }

    it('triggers when the edit-trigger is clicked', () => {
      const tree = testRender();

      enterEditMode(tree);

      expect(mockFocus).toHaveBeenCalledTimes(1);

      const input = tree.root.findByType('input');

      expect(input.props.value).toBe(defaultProps.value);
      expect(input.props.placeholder).toBe(defaultProps.placeholder);

      expect(tree).toMatchSnapshot();
    });

    it('exits when onBlur of the editContainer is called, calling onSubmit', () => {
      const tree = testRender();

      enterEditMode(tree);

      const editContainer = tree.root
        .findByProps({className: 'inline-editable__edit-container'});

      act(() => editContainer.props.onBlur());

      expect(mockOnSubmit).toHaveBeenCalledWith(defaultProps.value);

      expect(tree).toMatchSnapshot();
    });

    it('ignores non-escape key presses inside the editContainer', () => {
      const tree = testRender();

      enterEditMode(tree);

      const editContainer = tree.root
        .findByProps({className: 'inline-editable__edit-container'});

      act(() => editContainer.props.onKeyDown({key: 'Enter'}));

      expect(mockOnSubmit).toHaveBeenCalledTimes(0);

      expect(tree).toMatchSnapshot();
    });

    it('exits when the Escape key is pressed inside the editContainer, calling onSubmit', () => {
      const tree = testRender();

      enterEditMode(tree);

      const editContainer = tree.root
        .findByProps({className: 'inline-editable__edit-container'});

      act(() => editContainer.props.onKeyDown({key: 'Escape'}));

      expect(mockOnSubmit).toHaveBeenCalledWith(defaultProps.value);

      expect(tree).toMatchSnapshot();
    });
  });
});

