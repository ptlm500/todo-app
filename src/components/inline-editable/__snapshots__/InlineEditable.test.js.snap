// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineEditable edit mode exits when onBlur of the editContainer is called, calling onSubmit 1`] = `
<div
  className="inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      test value
    </span>
  </div>
</div>
`;

exports[`InlineEditable edit mode exits when the Escape key is pressed inside the editContainer, calling onSubmit 1`] = `
<div
  className="inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      test value
    </span>
  </div>
</div>
`;

exports[`InlineEditable edit mode ignores non-escape key presses inside the editContainer 1`] = `
<div
  className="inline-editable"
>
  <div
    className="inline-editable__edit-container"
    onBlur={[Function]}
    onKeyDown={[Function]}
  >
    <input
      placeholder="test placeholder"
      value="test value"
    />
  </div>
</div>
`;

exports[`InlineEditable edit mode triggers when the edit-trigger is clicked 1`] = `
<div
  className="inline-editable"
>
  <div
    className="inline-editable__edit-container"
    onBlur={[Function]}
    onKeyDown={[Function]}
  >
    <input
      placeholder="test placeholder"
      value="test value"
    />
  </div>
</div>
`;

exports[`InlineEditable renders a message to provide a placeholder when no value or placeholder is provided 1`] = `
<div
  className="inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      Provide a placeholder value
    </span>
  </div>
</div>
`;

exports[`InlineEditable renders a placeholder when no value is provided 1`] = `
<div
  className="inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      test placeholder
    </span>
  </div>
</div>
`;

exports[`InlineEditable renders a value when provided 1`] = `
<div
  className="inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      test value
    </span>
  </div>
</div>
`;

exports[`InlineEditable sets a className on the outer div when provided 1`] = `
<div
  className="test-class inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      test value
    </span>
  </div>
</div>
`;

exports[`InlineEditable throws an error when multiple child components are passed to it 1`] = `"[InlineEditable]: Multiple child components passed. InlineEditable only accepts one child"`;
